src/main/rx/fport.c

```diff
@@ -0,0 +1,282 @@
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "platform.h"
+
+#ifdef SERIAL_RX
+
+#include "build/debug.h"
+
+#include "common/utils.h"
+
+#include "drivers/time.h"
+
+#include "io/serial.h"
+
+#ifdef TELEMETRY
+#include "telemetry/telemetry.h"
+#endif
+
+#include "rx/rx.h"
+#include "rx/sbus_channels.h"
+#include "rx/fport.h"
+
+
+#define FPORT_TIME_NEEDED_PER_FRAME 5000
+
+#define FPORT_FRAME_PAYLOAD_SIZE_CONTROL (SBUS_CHANNEL_DATA_LENGTH + 3)
+#define FPORT_FRAME_PAYLOAD_SIZE_DOWNLINK 8
+#define FPORT_FRAME_MAX_PAYLOAD_SIZE FPORT_FRAME_PAYLOAD_SIZE_CONTROL 
+
+#define FPORT_FRAME_MARKER 0x7E
+
+#define FPORT_ESCAPE_CHAR 0x7D
+#define FPORT_ESCAPE_MASK 0x20
+
+#define FPORT_CRC_VALUE 0xFF
+
+#define FPORT_BAUDRATE 115200
+
+#define FPORT_PORT_OPTIONS (SERIAL_STOPBITS_1 | SERIAL_PARITY_NO)
+
+enum {
+	DEBUG_FPORT_FRAME_TYPE = 0,
+	DEBUG_FPORT_FRAME_INTERVAL,
+	DEBUG_FPORT_FRAME_ERRORS,
+	DEBUG_FPORT_FRAME_LAST_ERROR,
+};
+
+enum {
+	DEBUG_FPORT_NO_ERROR = 0,
+	DEBUG_FPORT_ERROR_OVERSIZE,
+	DEBUG_FPORT_ERROR_CHECKSUM,
+	DEBUG_FPORT_ERROR_FRAMING,
+	DEBUG_FPORT_ERROR_TIMEOUT,
+	DEBUG_FPORT_ERROR_PAYLOAD_SIZE = 0x10
+};
+
+typedef enum fportFrameState_e {
+	FPORT_FRAME_STATE_IDLE,
+	FPORT_FRAME_STATE_LENGTH,
+	FPORT_FRAME_STATE_PAYLOAD,
+	FPORT_FRAME_STATE_CHECKSUM,
+	FPORT_FRAME_STATE_END_MARKER,
+	FPORT_FRAME_STATE_COMPLETE,
+} fportFrameState_t;
+
+typedef enum fportFrameType_e {
+	FPORT_FRAME_TYPE_CONTROL = 0x00,
+	FPORT_FRAME_TYPE_DOWNLINK = 0x01,
+} fportFrameType_t;
+
+typedef struct fportControlData_s {
+	uint8_t sbusFrameStart;
+    sbusChannels_t channels;
+    uint8_t sbusFrameEnd;
+} fportControlData_t;
+
+typedef union fportData_s {
+	fportControlData_t controlData;
+} fportData_t;
+
+typedef struct fportFrameContents_s {
+    fportFrameType_t type;
+    fportData_t data;
+} fportFrameStructure_t;
+
+typedef union fportFramePayload_u {
+    uint8_t raw[FPORT_FRAME_MAX_PAYLOAD_SIZE];
+    fportFrameStructure_t structured;
+} fportFramePayload_t;
+
+typedef struct fportFrame_s {
+	uint8_t length;
+	fportFramePayload_t payload;
+} fportFrame_t;
+
+static volatile fportFrameState_t frameState = FPORT_FRAME_STATE_IDLE;
+static volatile uint16_t frameErrors = 0;
+
+static fportFrame_t fportFrame;
+
+static uint8_t calculateAdcChecksum(uint8_t checksum, const uint8_t val) {
+	checksum = checksum + val;
+	if (checksum < val) {
+		checksum++;
+	}
+
+	return checksum;
+}
+
+// Receive ISR callback
+static void fportDataReceive(uint16_t c)
```

Comment:
It may be better to use `FPORT_FRAME_MARKER` to decode frames
- stay idle until FPORT_FRAME_MARKER is received
- store all data (after unescaping, only checking buffer size) into buffer until next marker is seen
  - handle <FPORT_FRAME_MARKER> <FPORT_FRAME_MARKER> resonably
- process message on marker 
  - check it in ISR
  - (maybe preferable) let upper layer check frame and return to idle when done

Reply:
(BTW: why is frame size stored explicitly in frame? It is redundant with FPORT_FRAME_MARKER, the byte could be used to increase checksum size (providing much better error protection for both length and frame)

Reply:
Good point about post processing.
The frame marker / len duplicity (including the escaping business) was discussed with FrSky, but they insisted doing it this way, also with a view of at one point extending FPort to support a sensor bus a la SmartPort.

Reply:
(I don't see why sensor bus needs explicit length. IMO only cause is that bad design patterns from available code are reused ... )

Reply:
Will rework this to do decoding by `FPORT_FRAME_MARKER`.

Reply:
@ledvinap: I had a think about the `FPORT_FRAME_MARKER` vs. length field for frame matching situation. Since the 2 are redundant, I think it would make sense to use the length field to match frames. This way, we are able to mark the frame as complete and process it as soon as the last character has been received, and thus cut down control signal latency by one character's transmission time.
In addition to this, since there will always be one control frame directly followed by one downlink (sic) frame, this will buy us some time for the processing.

What are your thoughts on this?

Reply:
Using length only will sacrifice protection provided by FPORT_FRAME_MARKER and the gain is quite small.
It will also complicate decoding automaton slightly - you need new state for 'waiting for terminating frame marker'

Reply:
@ledvinap: Out of interest, what are the benefits of frame protection by marker over frame protection by length field? To me, both of these seem to be providing the same level of robustness.

I will probably move to a ring buffer design, so that follow up frames can be received while a frame is being processed.

Reply:
You get double protection when using both.

One advantage of using markers (with escaping) is frame length not limited to 8bit. 
Also decoding is deterministic - when you see frame marker, it must be start or end of frame (but that is related to start marker, not length). Receiving automaton can be simpler - just check for marker (and possibly address) and store data (checking receive buffer bounds). No header parsing etc.
Theoretically you can extend protocol to omit end marker for back-to-back (or fixed-length poll) packets, saving some bandwidth.

Cost for special char escaping is quite high (it will double packet in worst case). There are other methods to do this, for example https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing is quite simple and has favourable upper bound.

Reply:
@ledvinap: Thanks for the thorough analysis. I will take this into consideration when changing the code to use a ringbuffer. Because the protocol includes telemetry / data frames that are sent directly after the RC control frames, a ringbuffer design is needed to be able to receive frames while there are already frames queued up for processing.
Re frame markers / character escaping, this is a concern of mine too, but it's FrSky's design, not mine. Particularly for timing sensitive RC data, having a variable transport layer frame length seems to be a poor design choice.

Reply:
@mikeller : 
Using ringbuffer is probably overkill (you need frame boundaries / buffer wrap complicates things).
It may be better to allocate 2 (or few, it's easy to write generic code) packet buffers (large enough for packet) and implement packet queue/double buffering.

Variable length of frames is OK if there is enough time/saturation is correctly handled ... It's unlikely that someone does DOS attack ..